Settings changes

progwriter opened the pull request:
*Purpose*:
Partially addressing #894 
Allow settings to be copyable, which in turn allows task arguments to be layered on top of arguments passed to the worker at the start (as opposed to overriding them completely, which is what is happening currently and is not the desired behavior).

*Implementation*:
- Use the underlying configuration object for storing settings as opposed to local fields
- Keeping the existing getters/setter for compatibility
- Additional tests for settings
- Changes to `runBatfishThoughService`: task settings are copied over from `_mainSettings`

*Not in this PR*:
Settings hooks, i.e., allowing different plugins to register custom settings. Deferring to a different PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/batfish/batfish/956)
<!-- Reviewable:end -->

dhalperi commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 93 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jCVvB44C3MbXQsyem:-L5jCVvB44C3MbXQsyen:bf6l5tx) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L93)):*
> ```Java
>       String value = _line.getOptionValue(key);
>       boolean b;
>       if (value == null || value.trim().equalsIgnoreCase("true")) {
> ```

Add an implementation comment that we don't use `Boolean#parseBoolean` here because that does not do format detection?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 106 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jBNB5CSJQLAx5la9s:-L5jBNB5CSJQLAx5la9t:b-df8a82) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L106)):*
> ```Java
>   }
> 
>   public Configuration getConfiguration() {
> ```

Is this mutable? Can/should the caller mutate it?

Especially if mutable, but shouldn't, add to doc?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 112 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jC1qi4rGsRLL0wYpX:-L5jC1qj7KiayxSqq2oc:b-ig1var) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L112)):*
> ```Java
>   protected final Integer getIntegerOptionValue(String key) {
>     String valueStr = _line.getOptionValue(key);
>     if (valueStr == null) {
> ```

Can you flip the checks in this one, too, since you're flipping the others?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 115 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jC5696b0ilQonoOlk:-L5jC5696b0ilQonoOll:bp9v0tg) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L115)):*
> ```Java
>       return _config.getInteger(key, null);
>     } else {
>       return Integer.parseInt(valueStr);
> ```

So Integer.parseInt(" 1") throws an exception. Should we be trimming these like we trim booleans? Or should we be not trimming booleans? Or?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 135 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jGkSAlTBTSLzlgYu8:-L5jGkSAlTBTSLzlgYu9:buuzttu) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L135)):*
> ```Java
>     String valueStr = _line.getOptionValue(key);
>     if (valueStr != null) {
>       return Long.parseLong(valueStr);
> ```

Are these now re-parsed on every get?

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 156 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jDNvB0r1aYNzduEsp:-L5jDNvB0r1aYNzduEsq:b-eqxryo) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L156)):*
> ```Java
>     return s == null
>         ? Collections.emptyList()
>         : Arrays.stream(s).map(Paths::get).collect(Collectors.toList());
> ```

why not just store it as a list instead of an array? I see you can do this below.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 180 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jDr0b0TDyRGgNFI6k:-L5jDr0b0TDyRGgNFI6l:b60z2ae) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L180)):*
> ```Java
>   }
> 
>   protected final String getStringOptionValue(String key) {
> ```

Are these `@Nullable`? Are we guaranteed that `key` is present in the config for all uses?

---

*[projects/batfish/src/main/java/org/batfish/config/Settings.java, line 1374 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jGGbUBzCpioXmdPzv:-L5jGGbUBzCpioXmdPzw:bdy2jrd) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish/src/main/java/org/batfish/config/Settings.java#L1374)):*
> ```Java
> 
>   @VisibleForTesting
>   public void parseCommandLine(String[] args) {
> ```

This should only need to be `protected`, then.

---

*[projects/batfish/src/main/java/org/batfish/main/Driver.java, line 668 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jGS8-sbJw4ZyDPly7:-L5jGS8-sbJw4ZyDPly8:b4x8n2o) ([raw file](https://github.com/batfish/batfish/blob/d8c36ca20c360de826a28fef57af529f633084c7/projects/batfish/src/main/java/org/batfish/main/Driver.java#L668)):*
> ```Java
>           settings.setLogger(jobLogger);
> 
>           settings.setMaxRuntimeMs(_mainSettings.getMaxRuntimeMs());
> ```

Is this gone now? Unnecessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/956)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
*[projects/batfish/src/main/java/org/batfish/config/Settings.java, line 1374 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jGGbUBzCpioXmdPzv:-L5jHbP9cJrHVSX8WJdX:b-cusubh) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish/src/main/java/org/batfish/config/Settings.java#L1374)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

This should only need to be `protected`, then.
</blockquote></details>

Not even `protected`, just default visibility aka no modifier. Sorry.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/956)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 115 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jC5696b0ilQonoOlk:-L5jIWNBEizOdArhyEtg:buswv49) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L115)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

So Integer.parseInt(" 1") throws an exception. Should we be trimming these like we trim booleans? Or should we be not trimming booleans? Or?
</blockquote></details>

We probably shouldn't be trimming `boolean`s. I don't remember why we are.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 180 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jDr0b0TDyRGgNFI6k:-L5jIs3d395dfDZ6Ghtt:b-80r0rh) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L180)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Are these `@Nullable`? Are we guaranteed that `key` is present in the config for all uses?
</blockquote></details>

Expanding on that: if we don't use the old getters, is there now any way to distinguish between invalid configuration option and one that is default null?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/956)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 1 of 9 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/956)*
<!-- Sent from Reviewable.io -->

progwriter commented:
Review status: 7 of 10 files reviewed at latest revision, 17 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 93 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jCVvB44C3MbXQsyem:-L5jO_yPtLYeGY9BL1jV:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L93)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Add an implementation comment that we don't use `Boolean#parseBoolean` here because that does not do format detection?
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 106 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jBNB5CSJQLAx5la9s:-L5jR0RnUZvKwCbHtCeE:b66rmcl) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L106)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Is this mutable? Can/should the caller mutate it?

Especially if mutable, but shouldn't, add to doc?
</blockquote></details>

Removed all together for now as it's not strictly speaking necessary. When we add sub-settings from plugins I will think about the right way to do this.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 112 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jC1qi4rGsRLL0wYpX:-L5jQqZZ1by7IB8izyG_:b-jkgjy9) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L112)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Can you flip the checks in this one, too, since you're flipping the others?
</blockquote></details>

Thanks for catching this! Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 115 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jC5696b0ilQonoOlk:-L5jQtLLZzKphgOka2Li:b-ei603u) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L115)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

We probably shouldn't be trimming `boolean`s. I don't remember why we are.
</blockquote></details>

I was the one who added trims, but removed them now.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 135 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jGkSAlTBTSLzlgYu8:-L5jRBrxs-UC_z19MREG:biwrv8w) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L135)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Are these now re-parsed on every get?
</blockquote></details>

All of these `get*OptionValue` functions are only called during command line parsing. So yes, each time we parse the command line, but not any time after that.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 155 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jD_WC925AZYY_xzF6:-L5jRWxh_Y11SPBqRzOr:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L155)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Consider `ImmutableList.of()` for style, and for import minimization (if you use same import for non-empty lists as below).
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 156 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jDNvB0r1aYNzduEsp:-L5jRgBlQmhAZ_5pXlJj:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L156)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

why not just store it as a list instead of an array? I see you can do this below.
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 156 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jDSFd56H6CejEFqWf:-L5jS9g2IgA6y5KrCMqD:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L156)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Unless mutability is required, please use `ImmutableList.toImmutableList()` instead of `Collectors.toList()`.
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 172 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jEXg_A0tXNFOBun66:-L5jSVCflripFCH6HDUA:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L172)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

For style, `ImmutableList.of()`
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 174 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jEbEA3O4WWhJX4gDA:-L5jSVQuFvwMjyYA0pdq:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L174)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Unless mutability is required, use `ImmutableList.copyOf(Arrays.asList(optionValues))`.. or just `ImmutableList.of(optionValues)` if that exists.
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java, line 180 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jDr0b0TDyRGgNFI6k:-L5jVS1LSU3BnkKm9Yyr:b-wiq3pj) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish-common-protocol/src/main/java/org/batfish/common/BaseSettings.java#L180)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Expanding on that: if we don't use the old getters, is there now any way to distinguish between invalid configuration option and one that is default null?
</blockquote></details>

Few points to unpack here:
- At a high level, these can be `@Nullable` because there is no absolute guarantee key exists
- In our use cases, keys will exist because we initialize the defaults to some value prior to parsing the command line
- Eventually all of these can be made void, I have not done so just yet for compatibility with the `Settings` in coordinator and allinone
- An invalid configuration option (as passed on the command line) would throw an exception; Default or validly parsed values will go into the `_config`. At least thats my understanding of `CommandLine`

---

*[projects/batfish/src/main/java/org/batfish/config/Settings.java, line 577 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jGLlfjG4EVnhneQcf:-L5jSvpsUX6vETFVWez5:b-6yxe1f) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish/src/main/java/org/batfish/config/Settings.java#L577)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Why the default value?
</blockquote></details>

Because that's the default. Moved to the `initConfigDefaults`

---

*[projects/batfish/src/main/java/org/batfish/config/Settings.java, line 640 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jGj6jSWPzDXWxVE3_:-L5jSX3fY1MXbjRVAPBB:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish/src/main/java/org/batfish/config/Settings.java#L640)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Holy crap that's still here? I was going to make a comment about this line, but then realized this option is ancient (LogicBlox) and unused. The constant should be removed as well.
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/config/Settings.java, line 969 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jHL7RJlFOxNW5Q9SQ:-L5jTmVhYvYqawnoSkPS:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish/src/main/java/org/batfish/config/Settings.java#L969)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Why is there a default here instead of where defaults are set for everything else? Also use `ImmutableList.of()` for style.
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/config/Settings.java, line 1374 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jGGbUBzCpioXmdPzv:-L5jTtIFdccQZ8TZwTfH:bkfh2kb) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish/src/main/java/org/batfish/config/Settings.java#L1374)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Not even `protected`, just default visibility aka no modifier. Sorry.
</blockquote></details>

Ah, it should be public; `@VisibleForTesting` was an unnecessary decorator

---

*[projects/batfish/src/main/java/org/batfish/config/Settings.java, line 1586 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jI2ctqlJueBYcFmsn:-L5jU1-CNGyzPKyFV3U5:by48zq6) ([raw file](https://github.com/batfish/batfish/blob/10857e517ab183f6d292debea05eca057ea8fffe/projects/batfish/src/main/java/org/batfish/config/Settings.java#L1586)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

This seems weird. Why?
</blockquote></details>

It is weird. But `null` here serves a very special purpose (as I painstakingly discovered). A null path indicates the answer to a question should not be written to disk. So, to avoid a null pointer, we either store path as a string, or store `null`.

---

*[projects/batfish/src/main/java/org/batfish/main/Driver.java, line 668 at r1](https://reviewable.io:443/reviews/batfish/batfish/956#-L5jGS8-sbJw4ZyDPly7:-L5jSaYz2oG166CW7i_W:b-9jp1mj) ([raw file](https://github.com/batfish/batfish/blob/d8c36ca20c360de826a28fef57af529f633084c7/projects/batfish/src/main/java/org/batfish/main/Driver.java#L668)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Is this gone now? Unnecessary?
</blockquote></details>

Unnecessary as automatically copied from _mainSettings

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/956)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>  This is a great improvement of the structurally better type. Thanks.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/956#-:-L5ja5GTjTpzVCyRmrCx:big39ok)*
<!-- Sent from Reviewable.io -->

arifogel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/956#-:-L5kOfUCDYZc23GGNBiJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->
